### PR TITLE
Support for Freescale K64F MCU and FRDM-K64F Board

### PR DESCRIPTION
**Pull request by jfischer-phytec-iot**
This PR adds support for Freescale K64F MCU and FRDM-K64F Board.

~~base on #3099~~

If someone wants to test it, http://openocd.zylin.com/#/c/2773/ or
https://github.com/jfischer-phytec-iot/openocd/tree/wip%40phytec
could be relevant.

### CODE REVIEW

**Comment by jfischer-phytec-iot** [comment]
rebased on master, volunteers for review?

File: boards/frdm-k64f/Makefile.dep [hunk 1]
```diff
@@ -0,0 +1,4 @@
+ifneq (,$(filter ng_netif,$(USEMODULE)))
+  USEMODULE += kw2xrf
```
Comment (haukepetersen) [inline boards/frdm-k64f/Makefile.dep]: If I see it right, the board does natively not have a radio soldered on, right? So you should remove this file.

File: boards/frdm-k64f/Makefile.features [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+FEATURES_PROVIDED += periph_uart
+FEATURES_PROVIDED += periph_spi
+FEATURES_PROVIDED += periph_i2c
+FEATURES_PROVIDED += periph_pwm
+FEATURES_PROVIDED += periph_adc
+FEATURES_PROVIDED += periph_rtt
+FEATURES_PROVIDED += periph_rtc
+FEATURES_PROVIDED += periph_random
+FEATURES_PROVIDED += periph_cpuid
+FEATURES_MCU_GROUP = cortex_m4
```
Comment (haukepetersen) [inline boards/frdm-k64f/Makefile.features]: what about c++ support?

File: boards/frdm-k64f/Makefile.include [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# define the cpu used by the FRDM-K64F board
+export CPU = k64f
+export CPU_MODEL = mk64fn1m0vll12
+
+export MCPU = cortex-m4
```
Comment (haukepetersen) [inline boards/frdm-k64f/Makefile.include]: this is exported from the CPUs Makefile...

File: cpu/k64f/Makefile.include [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+export USEMODULE += kinetis_common
+
+# define path to kinetis module, which is needed for this CPU
+export KINETIS_COMMON = $(RIOTCPU)/kinetis_common/
+# CPU depends on the kinetis module, so include it
+include $(KINETIS_COMMON)Makefile.include
+
+export LINKFLAGS += -L$(RIOTCPU)/kinetis_common/ldscripts
+# define the linker script to use for this CPU
+export LINKERSCRIPT = $(CPU_MODEL).ld
```
Comment (haukepetersen) [inline cpu/k64f/Makefile.include]: I think you can drop this line, as the linkerscript is defined in `boards/Makefile.include.cortexm_comon`

File: cpu/k64f/Makefile.include [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+# tell the build system that the CPU depends on the Kinetis common files
+export USEMODULE += kinetis_common
+
+# define path to kinetis module, which is needed for this CPU
+export KINETIS_COMMON = $(RIOTCPU)/kinetis_common/
+# CPU depends on the kinetis module, so include it
+include $(KINETIS_COMMON)Makefile.include
+
+export LINKFLAGS += -L$(RIOTCPU)/kinetis_common/ldscripts
```
Comment (haukepetersen) [inline cpu/k64f/Makefile.include]: shouldn't this line better be defined in a `cpu/kinetis_common/Makefile.include`?

File: cpu/k64f/cpu.c [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Johann Fischer <j.fischer@phytec.de>
+ * @}
+ */
+
+#include <stdint.h>
+#include "cpu.h"
+#include "cpu_conf.h"
+
+#define FLASH_BASE          (0x00000000)
```
Comment (haukepetersen) [inline cpu/k64f/cpu.c]: not used -> remove?!

File: cpu/k64f/interrupt-vector.c [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+
+void dummy_handler(void)
+{
+    isr_unhandled();
+}
+
+/* Cortex-M specific interrupt vectors */
+void isr_svc(void)                  __attribute__((weak, alias("dummy_handler")));
+void isr_pendsv(void)               __attribute__((weak, alias("dummy_handler")));
+void isr_systick(void)              __attribute__((weak, alias("dummy_handler")));
```
Comment (haukepetersen) [inline cpu/k64f/interrupt-vector.c]: you could base your PR on #3155 and use the `WEAK_DEFAULT` macro here.

File: cpu/k64f/interrupt-vector.c [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+void isr_can0_wake_up(void)         __attribute__((weak, alias("dummy_handler")));
+void isr_sdhc(void)                 __attribute__((weak, alias("dummy_handler")));
+void isr_enet_1588_timer(void)      __attribute__((weak, alias("dummy_handler")));
+void isr_enet_transmit(void)        __attribute__((weak, alias("dummy_handler")));
+void isr_enet_receive(void)         __attribute__((weak, alias("dummy_handler")));
+void isr_enet_error(void)           __attribute__((weak, alias("dummy_handler")));
+
+
+/* interrupt vector table */
+__attribute__((section(".vector_table")))
```
Comment (haukepetersen) [inline cpu/k64f/interrupt-vector.c]: same here: you could use the `ISR_VECTORS` macro

**Comment by haukepetersen** [comment]
nice one! I think I might even have one of those boards in my office - will check on Monday. Besides some minor things, I think the main question is if we want to merge this PR first, or wait for #3155. I don't care...

**Comment by jfischer-phytec-iot** [comment]
@haukepetersen #3155 is more important for me, so i will rebase this PR on it. thx for review

File: cpu/k64f/Makefile.include (hunk 5 above)
Comment (jfischer-phytec-iot) [inline cpu/k64f/Makefile.include]: Actually yes, I will do it in a different PR for all Kinetis CPUs.

File: boards/frdm-k64f/Makefile.features (hunk 2 above)
Comment (jfischer-phytec-iot) [inline boards/frdm-k64f/Makefile.features]: should works

**Comment by haukepetersen** [comment]
ok, I do have the board at my office - though I am not there today... Will test tomorrow.

**Comment by jfischer-phytec-iot** [comment]
@haukepetersen You need https://github.com/jfischer-phytec-iot/openocd/tree/wip%40phytec (or  http://openocd.zylin.com/#/c/2773/) to test it.

**Comment by haukepetersen** [comment]
I was able to flash the board. After flashing, the RIOT startup output shows up once:
```
2015-06-19 15:41:39,944 - INFO # �kernel_init(): This is RIOT! (Version: 2014.12-1941-gb2989-HEAD)
2015-06-19 15:41:39,945 - INFO # kernel_init(): jumping into first task...
2015-06-19 15:41:39,945 - INFO # UART0 thread started.
2015-06-19 15:41:39,945 - INFO # uart0_init() [OK]
2015-06-19 15:41:39,946 - INFO # Welcome to RIOT!
2015-06-19 15:41:39,946 - INFO # > *** RIOT kernel panic
```
After that, the board does not give any other output (same when pressing reset). Can you reproduce this?

**Comment by haukepetersen** [comment]
wait, no idea what went wrong. After flashing again, dis- and reconnecting the board and then starting pyterm directly, the board is running the `default` example just fine.

**Comment by haukepetersen** [comment]
ok, the board seems to break after flashing. After calling `make flash term` it, it outputs some of the RIOT output to the terminal until it gets stuck somehow. When pressing the reset button there is no effect. But when pulling the USB connection and reconnecting it and opening a terminal (e.g. pyterm) agian, the output is just fine and the reset button does what it is supposed to do. Seems to me, like the CPU is in an undefined state or that the onboard UART-USB converter crashes or something. Any idea what this could be?

**Comment by jfischer-phytec-iot** [comment]
You need an update of CMSIS DAP firmware https://developer.mbed.org/handbook/Firmware-FRDM-K64F

**Comment by jfischer-phytec-iot** [comment]
And the mainline version of openocd does not work proper, see above.

**Comment by jfischer-phytec-iot** [comment]
@haukepetersen Are you still alive? :-)

**Comment by haukepetersen** [comment]
Yes, I am... I did use the patched OpenOCD version from you github fork. But I did not update the FRDM DAP firmware (yet). I will give this a try an let you know.

**Comment by haukepetersen** [comment]
Updated the DAP firmware (to v0221). Now I can't flash the device with OpenOCD anymore (still using the patched OpenOCD version).

I get
```
Info : CMSIS-DAP: SWD  Supported
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : CMSIS-DAP: FW Version = 1.0
Info : SWCLK/TCK = 0 SWDIO/TMS = 1 TDI = 0 TDO = 0 nTRST = 0 nRESET = 1
Info : CMSIS-DAP: Interface ready
Info : clock speed 1000 kHz
Info : SWD IDCODE 0x2ba01477
START...
Info : SWD IDCODE 0x2ba01477
Error: Failed to read memory at 0xe000ef40
Info : SWD IDCODE 0x2ba01477
Error: Failed to read memory at 0xe000ef44
Info : SWD IDCODE 0x2ba01477
Error: Failed to write memory at 0xe000edfc
Polling target k64f.cpu failed, trying to reexamine
START...
Info : SWD IDCODE 0x2ba01477
Info : SWD IDCODE 0x2ba01477
Error: Failed to read memory at 0xe000ed00
Examination failed, GDB will be halted. Polling again in 100ms
Polling target k64f.cpu failed, trying to reexamine
START...
Info : SWD IDCODE 0x2ba01477
Info : SWD IDCODE 0x2ba01477
Error: Failed to read memory at 0xe000ed00
Examination failed, GDB will be halted. Polling again in 300ms
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* k64f.cpu           cortex_m   little k64f.cpu           unknown
START...
Warn : *********** ATTENTION! ATTENTION! ATTENTION! ATTENTION! **********
Warn : ****                                                          ****
Warn : **** Your Kinetis MCU is in secured state, which means that,  ****
Warn : **** with exception for very basic communication, JTAG/SWD    ****
Warn : **** interface will NOT work. In order to restore its         ****
Warn : **** functionality please issue 'kinetis mdm mass_erase'      ****
Warn : **** command, power cycle the MCU and restart OpenOCD.        ****
Warn : ****                                                          ****
Warn : *********** ATTENTION! ATTENTION! ATTENTION! ATTENTION! **********
in procedure 'reset' 
in procedure 'ocd_bouncer'
```
I don't have the time right now to look further into it, will try to get to it soon.

**Comment by jfischer-phytec-iot** [comment]
@haukepetersen 
- Hold the reset button and execute ```make BOARD=frdm-k64f flash``` in a project directory
- Release reset button just after "Info : MDM: Chip is unsecured. Continuing."

**Comment by jfischer-phytec-iot** [comment]
rebased on master

**Comment by OlegHahm** [comment]
I cannot be ready for CI build and need squashing at the same time.

**Comment by haukepetersen** [comment]
@jfischer-phytec-iot: sorry for the delay... Flashed as you described, works like a charm now!

One thing I would like to ask you: could you put the information about the OpenOCD patches and the initial erase steps into the Wiki? I already started a rough page about the board [here](https://github.com/RIOT-OS/RIOT/wiki/Board:-FRDM-K64F). Thanks!

One more thing to add: ACK, please squash and add the 'Ready for CI build' tag.

**Comment by jfischer-phytec-iot** [comment]
@haukepetersen squashed, I also wrote a little...

**Comment by jfischer-phytec-iot** [comment]
The Travis was merciful :fireworks: .

**Comment by haukepetersen** [comment]
ACK and go!

**Comment by gebart** [comment]
Nice!

**Pull request merged**